core/src/main/scala/org/apache/spark/resource/ResourceProfile.scala

```diff
@@ -21,127 +21,160 @@ import java.util.{Map => JMap}
+    _id = ResourceProfile.DEFAULT_RESOURCE_PROFILE_ID
   }
 
-  def require(requests: TaskResourceRequests): this.type = {
-    _taskResources ++= requests.requests
-    this
+  override def equals(obj: Any): Boolean = {
+    obj match {
+      case that: ResourceProfile =>
+        that.getClass == this.getClass && that.id == _id &&
+          that.taskResources == taskResources && that.executorResources == executorResources
+      case _ =>
+        false
+    }
   }
 
+  override def hashCode(): Int = Seq(taskResources, executorResources).hashCode()
+
   override def toString(): String = {
-    s"Profile: id = ${_id}, executor resources: ${_executorResources}, " +
-      s"task resources: ${_taskResources}"
+    s"Profile: id = ${_id}, executor resources: ${executorResources.mkString(",")}, " +
+      s"task resources: ${taskResources.mkString(",")}"
   }
 }
 
-private[spark] object ResourceProfile extends Logging {
-  val UNKNOWN_RESOURCE_PROFILE_ID = -1
-  val DEFAULT_RESOURCE_PROFILE_ID = 0
-
+object ResourceProfile extends Logging {
+  // task resources
   val CPUS = "cpus"
+  // Executor resources
   val CORES = "cores"
   val MEMORY = "memory"
   val OVERHEAD_MEM = "memoryOverhead"
   val PYSPARK_MEM = "pyspark.memory"
 
+  // all supported spark executor resources (minus the custom resources like GPUs/FPGAs)
+  val allSupportedExecutorResources = Seq(CORES, MEMORY, OVERHEAD_MEM, PYSPARK_MEM)
+
+  val UNKNOWN_RESOURCE_PROFILE_ID = -1
+  val DEFAULT_RESOURCE_PROFILE_ID = 0
+
   private lazy val nextProfileId = new AtomicInteger(0)
 
   // The default resource profile uses the application level configs.
-  // Create the default profile immediately to get ID 0, its initialized later when fetched.
-  private val defaultProfileRef: AtomicReference[ResourceProfile] =
-    new AtomicReference[ResourceProfile](new ResourceProfile())
-
-  assert(defaultProfileRef.get().id == DEFAULT_RESOURCE_PROFILE_ID,
-    s"Default Profile must have the default profile id: $DEFAULT_RESOURCE_PROFILE_ID")
+  // var so that it can be reset for testing purposes.
+  private var defaultProfileRef: AtomicReference[ResourceProfile] =
+    new AtomicReference[ResourceProfile]()
 
-  def getNextProfileId: Int = nextProfileId.getAndIncrement()
+  private[spark] def getNextProfileId: Int = nextProfileId.getAndIncrement()
 
-  def getOrCreateDefaultProfile(conf: SparkConf): ResourceProfile = {
+  private[spark] def getOrCreateDefaultProfile(conf: SparkConf): ResourceProfile = {
     val defaultProf = defaultProfileRef.get()
     // check to see if the default profile was initialized yet
-    if (defaultProf.executorResources == Map.empty) {
-      synchronized {
-        val prof = defaultProfileRef.get()
-        if (prof.executorResources == Map.empty) {
-          addDefaultTaskResources(prof, conf)
-          addDefaultExecutorResources(prof, conf)
-        }
-        prof
+    if (defaultProf == null) {
+      val prof = synchronized {
```

Comment:
Aside from the previous comment, this doesn't seem exactly thread-safe, in the sense that it's still possible to create two "default" instances. Not sure how important that is.

(You could, instead of a synchronized, use `defaultProfileRef,compareAndSet`, and solve both problems.)

Reply:
Ok, now that I tried it again, I remember, I had tried compareAndSet at one point, but it always runs the code to generate the value to set when you do something like defaultProfileRef.compareAndSet(null, {function to generate value from application confs}). So even if its already set to not null it runs that code in the function.  I didn't want that.
 This is why I moved to just synchonrized block and it really doesn't matter if we create 2, only one of them will actually set it (which obviously I'm missing with your above comment).  I'll post a better version with it actually setting it and will be completely thread safe (just to be more clear) shortly.

Reply:
I went away from AtomicReference because this function isn't called very often anymore and I think its easier to read just using the lock.